<a href="https://colab.research.google.com/github/Wallace-NMelo/AM-CIN-UFPE/blob/main/AM_Desafio05_Wallace_N_Melo_02_asl.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<center><a href="https://www.nvidia.com/dli"> <img src="https://drive.google.com/uc?id=1J3bpGyJcz-7uOFkUNhvOBiReBCk-sUwR" width="200" </a></center>

# Image Classification of an American Sign Language Dataset

In this section, we will perform the data preparation, model creation, and model training steps we observed in the last section using a different dataset: images of hands making letters in [American Sign Language](http://www.asl.gs/).

## Objectives

* Prepare image data for training
* Create and compile a simple model for image classification
* Train an image classification model and observe the results

## The MNIST Dataset

In the history of deep learning, the accurate image classification of the [MNIST dataset](http://yann.lecun.com/exdb/mnist/), a collection of 70,000 grayscale images of handwritten digits from 0 to 9, was a major development. While today the problem is considered trivial, doing image classification with MNIST has become a kind of "Hello World" for deep learning.

In [ ]:
import numpy as np
import tensorflow as tf
from tensorflow import keras

In [ ]:
# Load the MNIST dataset
mnist = keras.datasets.mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()


11490434/11490434 [==============================] - 0s 0us/step


## Normalizing the data

Pixels values between 0 and 1 to facilitate classification

In [ ]:

train_images = train_images / 255.0
test_images = test_images / 255.0

This code defines a neural network with one hidden layer of 128 units and ReLU activation, and an output layer of 10 units with softmax activation to classify the 10 possible digits.

In [ ]:
# Define the neural network architecture
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(128, activation='relu'),
    keras.layers.Dense(10, activation='softmax')
])

The model is then compiled with the Adam optimizer and sparse categorical crossentropy loss function, and trained on the MNIST training set for 20 epochs.

In [ ]:
# Compile the model
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

In [ ]:
# Train the model
model.fit(train_images, train_labels, epochs=100)

# Evaluate the model on the test set
test_loss, test_acc = model.evaluate(test_images, test_labels)
print('Test accuracy:', test_acc)

Epoch 1/100
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0144 - accuracy: 0.9957
Epoch 2/100
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0122 - accuracy: 0.9963
Epoch 3/100
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0095 - accuracy: 0.9969
Epoch 4/100
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0094 - accuracy: 0.9972
Epoch 5/100
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0075 - accuracy: 0.9977
Epoch 6/100
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0082 - accuracy: 0.9973
Epoch 7/100
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0059 - accuracy: 0.9981
Epoch 8/100
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0057 - accuracy: 0.9982
Epoch 9/100
1875/1875 [==============================] - 5s 3ms/step - loss: 0.0060 - accuracy: 0.9982
Epoch 10/100
1875/1875 [==============================] - 5s 3ms/step - l

In [ ]:
# Evaluate the model on the test set
test_loss, test_acc = model.evaluate(test_images.reshape(-1, 28, 28, 1), test_labels)
print('Test accuracy:', test_acc)

313/313 [==============================] - 1s 2ms/step - loss: 0.2560 - accuracy: 0.9768
Test accuracy: 0.9768000245094299


The accuracy was between 97%~98% on the MNIST dataset.